sotodlib/io/bookbinder2.py

```diff
@@ -0,0 +1,1118 @@
+    # Look for zero-crossings
+    zc = []
+    while len(c_lower) > 0 and len(c_upper) > 0:
+        # Crossing from -ve to +ve
+        if c_lower[0] < c_upper[0]:
+            b = c_lower[c_lower < c_upper[0]]
+            zc.append(int( np.ceil(np.mean([b[-1], c_upper[0]])) ))
+            c_lower = c_lower[len(b):]
+        # Crossing from +ve to -ve
+        elif c_upper[0] < c_lower[0]:
+            b = c_upper[c_upper < c_lower[0]]
+            zc.append(int( np.ceil(np.mean([b[-1], c_lower[0]])) ))
+            c_upper = c_upper[len(b):]
+
+    # Replace all upper and lower crossings that contain a zero-crossing in
+    # between with the zero-crossing itself, but ONLY if those three events
+    # happen in quick succession (i.e., shorter than min_gap). Otherwise, they
+    # are separate events -- there is likely a stop state in between; in this
+    # case, do NOT perform the replacement.
+    for z in zc:
+        before_z = events[events < z]
+        after_z  = events[events > z]
+        if (after_z[0] - before_z[-1]) < min_gap:
+            events = np.concatenate((before_z[:-1], [z], after_z[1:]))
+
+    # Ignore first / last event if it is too close to the start or end
+    if (len(t) - events[-1] < min_gap) and events[-1] not in zc:
+        events = events[:-1]
+
+    if events[0] < min_gap:
+        events = events[1:]
+
+    return events + offset, starts + offset, stops + offset
+
+def find_frame_splits(ancil):
+    """
+    Determines timestamps of frame-splits from ACU data. If it cannot determine
+    frame-splits, returns None.
+    """
+    block = ancil.blocks['ACU_broadcast']
+    if 'Corrected_Azimuth' not in block.data:
+        return None 
+
+    az = block.data['Corrected_Azimuth']
+    idxs = locate_scan_events(az, filter_window=100)[0]
+    return block.times[idxs]
+
+
+def get_smurf_files(obs, meta_path, all_files=False):
+    """
+    Returns a list of smurf files that should be copied into a book.
+
+    Parameters
+    ------------
+    obs : G3tObservations
+        Observation to pull files from
+    meta_path : path
+        Smurf metadata path
+    all_files : bool
+        If true will return all found metadata files
+
+    Returns
+    -----------
+    files : List[path]
+        List of copyable files
+    """
+
+    def copy_to_book(file):
+        if all_files:
+            return True
+        return file.endswith('npy')
+
+    tscode = int(obs.action_ctime//1e5)
+    files = []
+
+    # check adjacent folders in case action falls on a boundary
+    for tc in [tscode-1, tscode, tscode + 1]:
+        action_dir = os.path.join(
+            meta_path,
+            str(tc),
+            obs.stream_id,
+            f'{obs.action_ctime}_{obs.action_name}'
+        )
+
+        if not os.path.exists(action_dir):
+            continue
+
+        for root, _, fs in os.walk(action_dir):
+            files.extend([os.path.join(root, f) for f in fs])
+
+    return [f for f in files if copy_to_book(f)]
```

Comment:
I like having newlines at the end of files but I suppose it isn’t strictly necessary. (Also applies to all the other files)